packages/curve-ui-kit/src/shared/ui/DataTable.tsx

```diff
@@ -0,0 +1,155 @@
+        paddingBlock: Spacing.md,
+        ...getExtraColumnPadding(column),
+      }}
+      data-testid={`data-table-cell-${column.id}`}
+    >
+      {flexRender(column.columnDef.cell, cell.getContext())}
+    </Typography>
+  )
+}
+
+const DataRow = <T extends unknown>({ row }: { row: Row<T> }) => {
+  const ref = useRef<HTMLTableRowElement>(null)
+  const entry = useIntersectionObserver(ref, { freezeOnceVisible: true })
+  return (
+    <TableRow
+      sx={{
+        marginBlock: 0,
+        height: Sizing['3xl'],
+        borderBottom: '1px solid',
+        borderColor: (t) => t.design.Layer[1].Outline,
+      }}
+      ref={ref}
+      // title={JSON.stringify(row.original)}
+      data-testid={`data-table-row-${row.id}`}
+    >
+      {entry?.isIntersecting && row.getVisibleCells().map((cell) => <DataCell key={cell.id} cell={cell} />)}
+    </TableRow>
+  )
+}
+
+/**
+ * In the figma design, the first and last columns seem to be aligned to the table title.
+ * However, the normal padding causes them to be misaligned.
+ */
+const getExtraColumnPadding = <T extends any>(column: Column<T>) => ({
+  ...(column.getIsFirstColumn() && { paddingInlineStart: Spacing.md }),
+  ...(column.getIsLastColumn() && { paddingInlineEnd: Spacing.md }),
+})
+
+const HeaderCell = <T extends unknown>({ header }: { header: Header<T, unknown> }) => {
+  const { column } = header
+  const sort = column.getIsSorted()
+  return (
+    <TableCell
+      sx={{
+        textAlign: getAlignment(column),
+        alignContent: 'end',
+        padding: Spacing.sm,
+        paddingBlockStart: 0,
+        ...getExtraColumnPadding(column),
+        cursor: 'pointer',
+      }}
+      colSpan={header.colSpan}
+      width={header.getSize()}
+      onClick={column.getToggleSortingHandler()}
+      data-testid={`data-table-header-${column.id}`}
+    >
+      <Typography variant="tableHeaderS" color={`text.${sort ? 'primary' : 'secondary'}`}>
+        {flexRender(column.columnDef.header, header.getContext())}
+        {sort && (
+          <ArrowDownIcon
+            sx={{ ...(sort === 'asc' && { transform: `rotate(180deg)` }), verticalAlign: 'text-bottom' }}
+          />
+        )}
+      </Typography>
+    </TableCell>
+  )
+}
+
+export const DataTable = <T extends unknown>({
+  columns,
+  data,
+  defaultSort,
+  headerHeight,
+}: {
+  data: T[]
+  defaultSort: SortingState
+  columns: ColumnDef<T, any>[]
+  headerHeight: string
+}) => {
+  const [sorting, onSortingChange] = useSortFromQueryString(defaultSort)
+  const table = useReactTable<T>({
+    columns,
+    data,
+    getCoreRowModel: getCoreRowModel(),
+    getSortedRowModel: getSortedRowModel(),
+    state: { sorting },
+    onSortingChange,
+    maxMultiSortColCount: 3, // allow 3 columns to be sorted at once
+  })
+  return (
+    <Table sx={{ minWidth: MinWidth.table }} data-testid="data-table">
+      <TableHead
+        sx={(t) => ({
+          zIndex: t.zIndex.appBar - 1,
+          position: 'sticky',
```

Comment:
I wonder how well this works if you have two tables on one page (possible stacked vertically)

Reply:
I think it should work fine, but if we ever need it we can customize

Reply:
[Screencast from 30-12-24 11:32:44.webm](https://github.com/user-attachments/assets/5a173e3a-2ec4-4f1b-b0b5-b489b81a99ba)